### PR TITLE
Add registry authentication

### LINKED ISSUE #86: Registry authentication
Hi, thanks for opensourcing this plugin.
I try to set it up with a private registry (gitlab), but I need to authenticate before pulling images. As I can see the auth config stanza is not supported.
Am I missing something ? Is it something doable ?

### PR DESCRIPTION
**Pull request by lisongmin**
fix #86 

I test pass locally.
But since the test required credential, how can i add a test for it?

@shishir-a412ed

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
@lisongmin Sorry for the delay, I ll take a look at this. Can you rebase?

**Comment by lisongmin** [comment]
> @lisongmin Sorry for the delay, I ll take a look at this. Can you rebase?

Done.

File: README.md [hunk 1]
```diff
@@ -111,6 +111,7 @@ More detailed instructions are in the [`example README.md`](https://github.com/R
 | **cap_add** | []string | no | Add individual capabilities. |
 | **cap_drop** | []string | no | Drop invidual capabilities. |
 | **devices** | []string | no | A list of devices to be exposed to the container. |
+| **auth** | block | no | Credential to get the image from a private repository. see [Auth](#auth) for detail. |
```
Comment (shishir-a412ed) [inline README.md:114]: Can we keep the description similar to `Nomad docker driver`?

```
`auth` - (Optional) Provide authentication for a private registry. See [Auth](#auth) for more details.
```

File: README.md [hunk 2]
```diff
@@ -162,6 +163,21 @@ config {
 }
 ```
 
+### auth
+
+To authenticate pulling private image, user can setting the `auth` stanza with `username` and `password` fields.
```
Comment (shishir-a412ed) [inline README.md:168]: Can we change this to:

```
If you want to pull from a private repository e.g. docker hub, you can specify `username` and `password` in the `auth` stanza. See example below.

**NOTE**: In the below example, `user` and `pass` are just placeholder values which need to be replaced by actual `username` and `password`, when specifying the credentials.
```

File: README.md [hunk 3]
```diff
@@ -162,6 +163,21 @@ config {
...
+```
+config {
+  auth {
+    username = "user"
+    password = "pass"
+  }
+}
+```
+
+if the credential is an identity token without `username`, set the identity token to the `password` field.
```
Comment (shishir-a412ed) [inline README.md:179]: nitpick: s/if/If

Comment (shishir-a412ed) [reply README.md:179]: I generated a docker access token for one of my private repositories (image) using https://docs.docker.com/docker-hub/access-tokens/
However, when I tried to run the nomad job with `auth` stanza, without specifying the `username` and just the `token`, it failed for me!

Does this work? Have you tried this with an example?

File: containerd/driver.go [hunk 4]
```diff
@@ -155,6 +159,12 @@ type Mount struct {
 	Options []string `codec:"options"`
 }
 
+// auth info to pull image from registry.
```
Comment (shishir-a412ed) [inline containerd/driver.go:162]: nitpick: s/auth/Auth

**Comment by shishir-a412ed** [comment]
@lisongmin 

Re: `But since the test required credential, how can i add a test for it?`
I am thinking, the happy case works, can we add a test for the negative case?

Can you add a test which tries to run a nomad job **without** `auth` stanza and tries to pull a private image `shm32/hello-world:private`.

This will fail with an error message containing:

```
pull access denied, repository does not exist or may require authorization
```

The test should `grep` for this message, and if it exists, the test should pass.

File: README.md (hunk 3 above)
Comment (lisongmin) [reply README.md:179]: we can see that dockhub needs `--username` with access-token in the link. so it should not work without a username.

I do not test this scene for i don't know how to get a idnetity token.  this scene is reference from https://github.com/containerd/containerd/blob/ab963e1cc16a845567a0e3e971775c29c701fcf8/pkg/cri/server/image_pull.go#L194

May be we should remove it?

Comment (shishir-a412ed) [reply README.md:179]: Yeah, let's remove this line for now, otherwise it might confuse the users.
We can add support for this (Authenticating just using the `identity token` without `username`) once we have a better way to test this scenario.

File: containerd/driver.go (hunk 4 above)
Comment (lisongmin) [reply containerd/driver.go]: Done

**Comment by lisongmin** [comment]
It seems the integration tests are not robust enough? I test pass locally. @shishir-a412ed 

```
vagrant@debian10:~/go/src/github.com/Roblox/nomad-driver-containerd$ ./tests/run_tests.sh tests/010-test-auth.sh
WARNING: Local host detected.
WARNING: These tests are designed to be run as part of continous integration (CI) and not recommended to be run on local host.
WARNING: These tests are destructive and can modify (or destroy) software on your host system.
WARNING: If running the tests locally, run it in the vagrant VM provided with the repository.
Do you still want to run the tests (Y/N)? y
INFO: Running tests on local host (vagrant VM). Setup is not required.
INFO: Checking if nomad-driver-containerd is up and running, and nomad is ready to accept jobs.
INFO: containerd driver is up and running.
INFO: Running test 010-test-auth.sh
~/go/src/github.com/Roblox/nomad-driver-containerd/example ~/go/src/github.com/Roblox/nomad-driver-containerd
INFO: Starting nomad auth job using nomad-driver-containerd.
==> Monitoring evaluation "2d262476"
    Evaluation triggered by job "auth"
    Evaluation within deployment: "bd552a4d"
    Allocation "fda3e22c" created: node "77bef1c1", group "auth-group"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "2d262476" finished with status "complete"
INFO: Checking can not pull image without auth info.
INFO: purge nomad auth job.
==> Monitoring evaluation "87d2c493"
    Evaluation triggered by job "auth"
    Evaluation within deployment: "bd552a4d"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "87d2c493" finished with status "complete"
~/go/src/github.com/Roblox/nomad-driver-containerd
PASS: 010-test-auth.sh
```

**Comment by shishir-a412ed** [comment]
@lisongmin 
I am seeing a different error now.

```
PASS: 009-test-allow-privileged.sh
496
INFO: Running test 010-test-auth.sh
497
~/go/src/github.com/Roblox/nomad-driver-containerd/example ~/go/src/github.com/Roblox/nomad-driver-containerd
498
INFO: Starting nomad auth job using nomad-driver-containerd.
499
Error submitting job: Put "http://127.0.0.1:4646/v1/jobs": dial tcp 127.0.0.1:4646: connect: connection refused
500
Error querying job: Get "http://127.0.0.1:4646/v1/jobs?prefix=auth": dial tcp 127.0.0.1:4646: connect: connection refused
```
Digging a bit more. Will update here.

**Comment by shishir-a412ed** [comment]
@lisongmin So the problem is, `009-test-allow-privileged.sh` should always be the **last** test to run. This test requires a  `nomad` restart and any tests that will run after this, may or may not fail. I ll probably document this in the [`Tests`](https://github.com/Roblox/nomad-driver-containerd#tests) section. 

For now,

Can you move `auth` test to `009-test-auth.sh` and `allow-privileged` test to `010-test-allow-privileged.sh`?

File: README.md [hunk 5]
```diff
@@ -111,6 +111,7 @@ More detailed instructions are in the [`example README.md`](https://github.com/R
 | **cap_add** | []string | no | Add individual capabilities. |
 | **cap_drop** | []string | no | Drop invidual capabilities. |
 | **devices** | []string | no | A list of devices to be exposed to the container. |
+| **auth** | block | no | (Optional) Provide authentication for a private registry. See [Auth](#auth) for more details. |
```
Comment (shishir-a412ed) [inline README.md:114]: nitpick: Remove `(Optional)`

File: example/auth.nomad [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  group "auth-group" {
+    task "auth-task" {
+      driver = "containerd-driver"
+
+      config {
+        image = "shm32/hello-world:private"
+        entrypoint = ["/bin/sleep"]
+        args = ["600s"]
```
Comment (shishir-a412ed) [inline example/auth.nomad:17]: Do we need the `entrypoint` and `args` here?

**Review by shishir-a412ed — APPROVED** [review]
LGTM!
